mockk/src/main/kotlin/io/mockk/impl/annotations/JvmMockInitializer.kt

```diff
@@ -66,6 +80,38 @@ class JvmMockInitializer(val gateway: MockKGateway) : MockKGateway.MockInitializ
         }
     }
 
+    private fun matchInjectMockMembers(target: Any, property: KProperty1<Any, Any>) {
```

Comment:
Overall I think we should first learn how Mockito implements this feature. And probably do even better ;-)  

Mockito ***creates*** an instance marked with InjectMocks substituting already created mocks to a constructor. 

Resolution rules are an important part of the design for this task. It should be pretty clear what is happening in case of two Mocks of same types, e.t.c.

Reply:
Thanks for reviewing this PR, didn't want to proceed further until you did a quick check.

Definitely agree on being better than Mockito with the `@InjectMocks` implementation.

`@InjectMocks` seems to be problematic for many with Mockito.  The main problems you'll find when googling Mockito's `@InjectMocks` is how unpredictable their behavior is between same types defined in the test vs the injectable mock target.  They try to match by the property name or the string argument given to `@Mock` annotation (i.e. `@Mock("foo")`).  If they fail, they don't report errors out.  We should definitely be clear on the resolution rules so users do not see unpredictable behavior of same types.  We can also be better with the error handling.

Based on the [documentation](http://static.javadoc.io/org.mockito/mockito-core/2.15.0/org/mockito/InjectMocks.html), Mockito can create the `@InjectMock` property either 2 ways.
- no-arg constructor
- constructor with most parameters (used during construction strategy as opposed to setter or property inject strategies)

If construction strategy fails, setter or property strategy is explored by first creating the instance with zero-arg constructor.  Documentation linked above has good information.